views/dm/characters.ejs

```diff
@@ -21,146 +47,148 @@
-                        switch (snapshot.state) {
-                            case firebase.storage.TaskState.PAUSED: // or 'paused'
-                                console.log('Upload is paused');
-                                break;
-                            case firebase.storage.TaskState.RUNNING: // or 'running'
-                                console.log('Upload is running');
-                                break;
-                        }
-                    },
-                    (error) => {
-                        reject(err);
-                    },
-                    () => {
-                        // Upload completed successfully, now we can get the download URL
-                        uploadTask.snapshot.ref.getDownloadURL().then((downloadURL) => {
-                            modalCharacterForm.find(`#characterForm #${pictureFieldId}`).val(downloadURL);
-                            resolve();
-                        });
-                    }
+        <script src="/javascripts/MapHelper.js"></script>
+        <script src="/javascripts/CharacterHelper.js"></script>
+        <script>
+            let container;
+            let characters = [];
+            let modalCharacterForm;
+            let modalCharacterFooterButtons;
+            let currentCharacterFormUrl = getCharacterFormUrl(roomKey);
+            const hourglassIcon = `<i class="ra ra-hourglass" id="form-loading-icon"></i>`;
+
+            function uuidv4() {
+                return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, c =>
+                    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
                 );
-            });
-        }
-
-        $(() => {
-            modalCharacterForm = $('#characterFormModal');
-            modalCharacterFooterButtons = modalCharacterForm.find('.modal-footer button');
-            container = $('#characters-container');
-
-            container.on('click', '.edit-btn', (e) => {
-                const characterId = $(e.currentTarget).attr('data-fs-id');
-                currentCharacterFormUrl = getCharacterFormUrl(roomKey, characterId);
-                modalCharacterForm.modal('show');
-            })
+            }
+
+            function addCharacter(character) {
+                characters.push(character);
+                firestore
+                    .collection('rooms')
+                    .doc(roomKey)
+                    .collection('characters')
+                    .add(character)
+                renderCharacters();
+            }
+
+            function renderCharacters() {
+                container.text('');
+                characters.forEach((character) => {
+                    container.append(CharacterHelper.generateCharacterHTML(character));
+                    util_createImageRenderDom(character.id, character.Picture) // Adds images to the dom hidden so we can use them later
```

Comment:
Is this necessary since you should be reusing the image from the character profile?

Reply:
Noted. Will try to grab it from the existing DOM element. If that works, I'll nuke my function that adds the picture into a new element.

Reply:
Resolved that in 2890b85c1d48a0bbc5ffca457fb45675110ea3f5